storage/pebble/execution_data_tracker.go

```diff
@@ -0,0 +1,406 @@
+func (s *ExecutionDataTracker) trackBlob(blockHeight uint64, c cid.Cid) error {
+	err := operation.InsertBlob(blockHeight, c)(s.db)
+	if err != nil {
+		return fmt.Errorf("failed to add blob record: %w", err)
+	}
+
+	var latestHeight uint64
+	err = operation.RetrieveTrackerLatestHeight(c, &latestHeight)(s.db)
+	if err != nil {
+		if !errors.Is(err, storage.ErrNotFound) {
+			return fmt.Errorf("failed to get latest height: %w", err)
+		}
+	} else {
+		// don't update the latest height if there is already a higher block height containing this blob
+		if latestHeight >= blockHeight {
+			return nil
+		}
+	}
+
+	err = operation.UpsertTrackerLatestHeight(c, blockHeight)(s.db)
+	if err != nil {
+		return fmt.Errorf("failed to set latest height value: %w", err)
+	}
+
+	return nil
+}
+
+// trackBlobs tracks multiple blob CIDs at the specified block height.
+//
+// Parameters:
+// - blockHeight: The height at which the blobs were observed.
+// - cids: The CIDs of the blobs to be tracked.
+//
+// No errors are expected during normal operation.
+func (s *ExecutionDataTracker) trackBlobs(blockHeight uint64, cids ...cid.Cid) error {
+	if len(cids) > 0 {
+		for _, c := range cids {
+			if err := s.trackBlob(blockHeight, c); err != nil {
+				return fmt.Errorf("failed to track blob %s: %w", c.String(), err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// batchDelete deletes multiple blobs from the storage in a batch operation.
+//
+// Parameters:
+// - deleteInfos: Information about the blobs to be deleted, including their heights and
+// whether to delete the latest height record.
+//
+// No errors are expected during normal operation.
+func (s *ExecutionDataTracker) batchDelete(deleteInfos []*storage.DeleteInfo) error {
+	for _, dInfo := range deleteInfos {
+		err := operation.RemoveBlob(dInfo.Height, dInfo.Cid)(s.db)
+		if err != nil {
+			return fmt.Errorf("failed to delete blob record for Cid %s: %w", dInfo.Cid.String(), err)
+		}
+
+		if dInfo.DeleteLatestHeightRecord {
+			err = operation.RemoveTrackerLatestHeight(dInfo.Cid)(s.db)
+			if err != nil {
+				return fmt.Errorf("failed to delete latest height record for Cid %s: %w", dInfo.Cid.String(), err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// PruneUpToHeight removes all data from storage corresponding
+// to block heights up to and including the given height,
+// and updates the latest pruned height value.
+// It locks the ExecutionDataTracker and ensures that no other writes
+// can occur during the pruning.
+// It is up to the caller to ensure that this is never
+// called with a value higher than the fulfilled height.
+//
+// No errors are expected during normal operation.
+func (s *ExecutionDataTracker) PruneUpToHeight(height uint64) error {
+	blobRecordPrefix := []byte{storage.PrefixBlobRecord}
+	itemsPerBatch := storage.DeleteItemsPerBatch
+	var batch []*storage.DeleteInfo
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if err := s.setPrunedHeight(height); err != nil {
```

Comment:
I think we should set this set this at the end as part of the batch. it's not a huge deal since the next prune cycle would delete data that's supposed to be pruned, but the getter may return the incorrect height